[AutoDiff] Revamp usefulness propagation in activity analysis.

dan-zheng opened the pull request:
Useful values are those that contribute to (specific) dependent variables,
i.e. function results.

For addresses: all projections of a useful address should be useful.
This has special support:
`DifferentiableActivityInfo::propagateUsefulThroughAddress`.

Previously:
- Usefulness was propagated by iterating through all instructions in
  post-dominance order. This is not efficient because irrelevant instructions
  may be visited.
- For useful addresses, `propagateUsefulThroughAddress` propagated
  usefulness one step to projections, but not recursively to users of the
  projections. This caused some values to incorrectly not be marked useful.

Now:
- Usefulness is propagated by following use-def chains, starting from
  dependent variables (function results). This is handled by the
  following helpers:
  - `setUsefulAndPropagateToOperands(SILValue, unsigned)`: marks a value as
    useful and recursively propagates usefulness through defining instruction
    operands and basic block argument incoming values.
  - `propagateUseful(SILInstruction *inst, unsigned)`: propagates usefulness
    to the operands of the given instruction.
- `DifferentiableActivityInfo::propagateUsefulThroughAddress` now calls
  `propagateUseful` to propagate usefulness recursively through users'
  operands.

Effects:
- More values are now (correctly) marked as useful, affecting
  non-differentiability diagnostics for active enum values (TF-956) and for-in
  loops (TF-957). Both have room for improvement.

Resolves control flow differentiation correctness issue: TF-954.

dan-zheng commented:
Created PR as a draft because I feel there's room for improvement.
Suggestions/feedback are welcome!

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1444,8 +1444,18 @@ class DifferentiableActivityInfo {
   void setUseful(SILValue value, unsigned dependentVariableIndex);
```
Note: currently, `setUseful` has two users (`setUsefulAndPropagateToOperands` and `propagateUsefulThroughAddress`), so it hasn't been inlined.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1975,6 +1985,71 @@ void DifferentiableActivityInfo::propagateVaried(
+  auto *inst = value->getDefiningInstruction();
+  if (!inst)
+    return;
+  propagateUseful(inst, dependentVariableIndex);
```
Note: `propagateUseful` cannot be inlined because it has multiple users (called multiple times in `propagateUsefulThroughAddress`.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1975,6 +1985,71 @@ void DifferentiableActivityInfo::propagateVaried(
+    propagateUsefulThroughAddress(value, dependentVariableIndex);
+    return;
+  }
+  setUseful(value, dependentVariableIndex);
```
Note: both `setUsefulAndPropagateToOperands` and `propagateUsefulThroughAddress` call `isUseful` and `setUseful`, which seems suboptimal. I tried removing `setUseful` from one of them but ran into infinite loops.

Related: I think `setUsefulAndPropagateToOperands` should be the primary entry point for propagating usefulness, so I eliminated most direct calls to `propagateUsefulThroughAddress`.

rxwei commented on test/AutoDiff/activity_analysis.swift:
```diff
@@ -88,7 +88,7 @@ func testArrayUninitializedIntrinsicGeneric<T>(_ x: T, _ y: T) -> [T] {
 // CHECK: [ACTIVE]   %6 = apply %5<T>(%4) : $@convention(thin) <τ_0_0> (Builtin.Word) -> (@owned Array<τ_0_0>, Builtin.RawPointer)
 // CHECK: [ACTIVE] (**%7**, %8) = destructure_tuple %6 : $(Array<T>, Builtin.RawPointer)
-// CHECK: [VARIED] (%7, **%8**) = destructure_tuple %6 : $(Array<T>, Builtin.RawPointer)
+// CHECK: [ACTIVE] (%7, **%8**) = destructure_tuple %6 : $(Array<T>, Builtin.RawPointer)
```
This now doesn't quite make sense to me: Why is `Builtin.RawPointer` marked active? The previous "set activity across array initialization" logic was working fine and it would make `%8` have `[NONE]` activity.

dan-zheng replied:
Undid the change in https://github.com/apple/swift/pull/28225/commits/86a7b97c3dae302c1014eb2970f6b6f2290f9499. I found that propagating usefulness through the `RawPointer` returned by `array.uninitialized_intrinsic` applications fixes a version of TF-952, but I can explore that in a follow-up.

dan-zheng committed:
[AutoDiff] Revamp usefulness propagation in activity analysis.

Useful values are those that contribute to (specific) dependent variables,
i.e. function results.

For addresses: all projections of a useful address should be useful.
This has special support:
`DifferentiableActivityInfo::propagateUsefulThroughAddress`.

Previously:
- Usefulness was propagated by iterating through all instructions in
  post-dominance order. This is not efficient because irrelevant instructions
  may be visited.
- For useful addresses, `propagateUsefulThroughAddress` propagated
  usefulness one step to projections, but not recursively to users of the
  projections. This caused some values to incorrectly not be marked useful.

Now:
- Usefulness is propagated by following use-def chains, starting from
  dependent variables (function results). This is handled by the
  following helpers:
  - `setUsefulAndPropagateToOperands(SILValue, unsigned)`: marks a value as
    useful and recursively propagates usefulness through defining instruction
    operands and basic block argument incoming values.
  - `propagateUseful(SILInstruction *inst, unsigned)`: propagates usefulness
    to the operands of the given instruction.
- `DifferentiableActivityInfo::propagateUsefulThroughAddress` now calls
  `propagateUseful` to propagate usefulness recursively through users'
  operands.

Effects:
- More values are now (correctly) marked as useful, affecting
  non-differentiability diagnostics for active enum values (TF-956) and for-in
  loops (TF-957). Both have room for improvement.

Resolves control flow differentiation correctness issue: TF-954.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented:
@swift-ci Please test tensorflow

rxwei commented:
@swift-ci Please test tensorflow

rxwei commented:
@swift-ci Please test tensorflow
